vochain/app.go

```diff
@@ -224,84 +215,145 @@ func (app *BaseApplication) InitChain(req abcitypes.RequestInitChain) abcitypes.
 	if err := app.State.SetElectionPriceCalc(); err != nil {
-		log.Fatalf("cannot set election price calc: %v", err)
+		return nil, fmt.Errorf("cannot set election price calc: %w", err)
 	}
 
 	// commit state and get hash
 	hash, err := app.State.Save()
 	if err != nil {
-		log.Fatalf("cannot save state: %s", err)
+		return nil, fmt.Errorf("cannot save state: %w", err)
 	}
-	return abcitypes.ResponseInitChain{
+	return &abcitypes.ResponseInitChain{
 		Validators: tendermintValidators,
 		AppHash:    hash,
-	}
+	}, nil
 }
 
 // CheckTx unmarshals req.Tx and checks its validity
-func (app *BaseApplication) CheckTx(req abcitypes.RequestCheckTx) abcitypes.ResponseCheckTx {
-	height := app.Height()
+func (app *BaseApplication) CheckTx(ctx context.Context,
+	req *abcitypes.RequestCheckTx) (*abcitypes.ResponseCheckTx, error) {
 	if req.Type == abcitypes.CheckTxType_Recheck {
-		if height%recheckTxHeightInterval != 0 {
-			return abcitypes.ResponseCheckTx{Code: 0}
+		if app.Height()%recheckTxHeightInterval != 0 {
+			return &abcitypes.ResponseCheckTx{Code: 0}, nil
 		}
 	}
 	tx := new(vochaintx.Tx)
 	if err := tx.Unmarshal(req.Tx, app.ChainID()); err != nil {
-		return abcitypes.ResponseCheckTx{Code: 1, Data: []byte("unmarshalTx " + err.Error())}
+		return &abcitypes.ResponseCheckTx{Code: 1, Data: []byte("unmarshalTx " + err.Error())}, err
 	}
 	response, err := app.TransactionHandler.CheckTx(tx, false)
 	if err != nil {
 		if errors.Is(err, transaction.ErrorAlreadyExistInCache) {
-			return abcitypes.ResponseCheckTx{Code: 0}
+			return &abcitypes.ResponseCheckTx{Code: 0}, nil
 		}
 		log.Errorw(err, "checkTx")
-		return abcitypes.ResponseCheckTx{Code: 1, Data: []byte("checkTx " + err.Error())}
+		return &abcitypes.ResponseCheckTx{Code: 1, Data: []byte("checkTx " + err.Error())}, err
 	}
-	return abcitypes.ResponseCheckTx{
+	return &abcitypes.ResponseCheckTx{
 		Code: 0,
 		Data: response.Data,
 		Info: fmt.Sprintf("%x", response.TxHash),
 		Log:  response.Log,
+	}, nil
+}
+
+// FinalizeBlock It delivers a decided block to the Application. The Application must execute
+// the transactions in the block deterministically and update its state accordingly.
+// Cryptographic commitments to the block and transaction results, returned via the corresponding
+// parameters in ResponseFinalizeBlock, are included in the header of the next block.
+// CometBFT calls it when a new block is decided.
+func (app *BaseApplication) FinalizeBlock(ctx context.Context,
+	req *abcitypes.RequestFinalizeBlock) (*abcitypes.ResponseFinalizeBlock, error) {
+	height := uint32(req.GetHeight())
+	app.beginBlock(req.GetTime(), height)
+	txResults := make([]*abcitypes.ExecTxResult, len(req.Txs))
+	for i, tx := range req.Txs {
+		resp := app.deliverTx(tx)
+		if resp.Code != 0 {
+			log.Warnw("deliverTx failed",
+				"code", resp.Code,
+				"data", string(resp.Data),
+				"info", resp.Info,
+				"log", resp.Log)
+		}
+		txResults[i] = &abcitypes.ExecTxResult{
```

Comment:
should we include all txs even if the response code of the deliverTx is != 0 ?

Reply:
this is just the result of the execution. So, you need to include a result even if the code is different to 0